Better_Cron_Estimates

ExtremeFiretop opened the pull request:
Better_Cron_Estimates

ExtremeFiretop commented:
This correcly factors in the cron runtime to the estimates

But I haven't added the translation for Days or Months as text instead of numbers.
i.e (Sun or Mar) wouldn't work with this method yet.

But this method does correctly estimate the next run time factoring the delay and cron schedule.
Only thing left I think is to add the conversion table for the numbers to text.

Any thoughts on this? No pressure not trying to push any specific way.
Just trying to deal with Viktors suggestion of better estimation since our current method only factors the postpone time and not the cron schedule.

ExtremeFiretop commented:
Funny enough another request for something like this just came in:

https://www.snbforums.com/threads/merlinau-v1-1-2-the-ultimate-firmware-auto-updater-now-available-in-amtm.88577/post-905910

ExtremeFiretop commented:
Added the conversion map.
Seems to work correctly as found below:

![image](https://github.com/ExtremeFiretop/MerlinAutoUpdate-Router/assets/1971404/96295549-59a8-4fd7-827d-2269f407410b)

![image](https://github.com/ExtremeFiretop/MerlinAutoUpdate-Router/assets/1971404/835b19ae-040b-4e87-a7fd-040c0e3254d3)

![image](https://github.com/ExtremeFiretop/MerlinAutoUpdate-Router/assets/1971404/85287862-fa4d-4805-87db-612daa18d818)

![image](https://github.com/ExtremeFiretop/MerlinAutoUpdate-Router/assets/1971404/ebb25c60-c3a6-4d86-b1f5-fcc9e0c46325)

ExtremeFiretop commented:
@Martinski4GitHub 

Added the missing pieces after I finished testing everything and it all appears to be working but I'll let you nitpick it apart :)

Martinski4GitHub commented:
> 
> Added the missing pieces after I finished testing everything and it all appears to be working but I'll let you nitpick it apart :) Your feedback is really appreciated as you may notice something while testing that I didn't
>
I'll try to review/test it later this evening if I can; otherwise, tomorrow (we have friends coming over for dinner in about half-hour so most likely will be tomorrow). But I just gave it a quick look and it's looking good.

Martinski4GitHub commented:
I found one problem where if the cron job schedule cannot be parsed with the new code, the message simply says:

`No suitable date found for the firmware update within the next year.`

![RT-AC86U_FWUpdateExpected1](https://github.com/ExtremeFiretop/MerlinAutoUpdate-Router/assets/119833648/68e4743e-49e5-41a6-aa7a-d6545c1be429)

In such cases where the cron schedule cannot be parsed, the code should fall back to the "old" message so that at least the user gets some indication of when to expect the F/W update to occur; otherwise, he's now left in the "dark" and perhaps thinking that something went wrong or is currently wrong with the cron job.

![RT-AC86U_FWUpdateExpected3](https://github.com/ExtremeFiretop/MerlinAutoUpdate-Router/assets/119833648/2821a91c-e29d-45ed-ac37-b85f21a89c4f)

ExtremeFiretop commented:
> > Added the missing pieces after I finished testing everything and it all appears to be working but I'll let you nitpick it apart :) Your feedback is really appreciated as you may notice something while testing that I didn't
> 
> I'll try to review/test it later this evening if I can; otherwise, tomorrow (we have friends coming over for dinner in about half-hour so most likely will be tomorrow). But I just gave it a quick look and it's looking good.

Thanks buddy take your time I need to review your PR now too! :)

ExtremeFiretop commented:
> I found one problem where if the cron job schedule cannot be parsed with the new code, the message simply says:
> 
> `No suitable date found for the firmware update within the next year.`
> 
> ![RT-AC86U_FWUpdateExpected1](https://private-user-images.githubusercontent.com/119833648/327985562-68e4743e-49e5-41a6-aa7a-d6545c1be429.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQ4ODAwOTIsIm5iZiI6MTcxNDg3OTc5MiwicGF0aCI6Ii8xMTk4MzM2NDgvMzI3OTg1NTYyLTY4ZTQ3NDNlLTQ5ZTUtNDFhNi1hYTdhLWQ2NTQ1YzFiZTQyOS5qcGc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNTA1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDUwNVQwMzI5NTJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01NGU5ZDM2YTNiMDE1YjZjMzNiNzIyZTQ0OWZlODRmYmI3YWI0ZGUwNjJjYTE2NmI4MDk2NTgzNDk4NjZmNTE1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.MM3qjv5gLsPj4cBznfDnIkWr39OsLI5M-lsU2FpY3rM)
> 
> In such cases where the cron schedule cannot be parsed, the code should fall back to the "old" message so that at least the user gets some indication of when to expect the F/W update to occur; otherwise, he's now left in the "dark" and perhaps thinking that something went wrong or is currently wrong with the cron job.
> 
> ![RT-AC86U_FWUpdateExpected3](https://private-user-images.githubusercontent.com/119833648/327985993-2821a91c-e29d-45ed-ac37-b85f21a89c4f.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQ4ODAwOTIsIm5iZiI6MTcxNDg3OTc5MiwicGF0aCI6Ii8xMTk4MzM2NDgvMzI3OTg1OTkzLTI4MjFhOTFjLWUyOWQtNDVlZC1hYzM3LWI4NWYyMWE4OWM0Zi5qcGc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNTA1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDUwNVQwMzI5NTJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kZTAyMmNlOWI0NjQ3YWMwYzBjZjlkZWE1NTZjY2M5ZDA4NmJhODRkMjZkN2YwZjkzNGM4MDM4ZGUyYTUxOTg4JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.9FLCVj6EmhrZFugE03iB2GDb3OXPRd0V554CJWJyVFs)

Ha!

This is exactly why I need another set of eyes. Thank you @Martinski4GitHub 
You da bestest! <3

ExtremeFiretop commented:
@Martinski4GitHub 

Addressed both feedback in commit: https://github.com/ExtremeFiretop/MerlinAutoUpdate-Router/pull/205/commits/0e07f8be93d23652a5a9f6a9ce58b84e3b069a4d

![image](https://github.com/ExtremeFiretop/MerlinAutoUpdate-Router/assets/1971404/ed80ad29-51ff-4ea3-84ef-5055f53fcf2e)

1. It now falls back to the old method if it can't find a future date matching the cron within a year.
2. It can properly handle multiple Days in the cron.

However I noticed another bug now I need to quickly work out.
If I have it set as "May" instead for the cron it gives me a estimate in the new year:

![image](https://github.com/ExtremeFiretop/MerlinAutoUpdate-Router/assets/1971404/4f12c9a9-7ded-4a88-9356-865ab582c435)

But it should be sometime this month surely.

ExtremeFiretop commented:
@Martinski4GitHub 

I added the fallback if a future date isn't found.

There seems to still be 2 bugs:

1. If you set a month in the cron, for example 20 5 * 5 Sun (month being May)
It for some reason won't pickup the recent match in may 2024, but will for 2025 lol

![image](https://github.com/ExtremeFiretop/MerlinAutoUpdate-Router/assets/1971404/57a65f21-bb38-4c2a-bc06-059980c528c8)

2. The function cannot intake more than one value at a time. 
Crons such as **20 5 * 5 Sat,Sun** Will never find a matching future date.

ExtremeFiretop commented:
More investigation, but I'll do that tomorrow my head hurts and I can't follow the code anymore.
If you have any ideas what is causing this feel free, otherwise I'll start fresh tomorrow.

ExtremeFiretop commented:
Okay, fixed but 1.

Working on bug 2 next.

ExtremeFiretop commented:
It can now handle and translate multiple days and months:

![image](https://github.com/ExtremeFiretop/MerlinAutoUpdate-Router/assets/1971404/3fb6e928-4107-4555-836c-c0d532ea3f56)

And it now properly matches when a matching date is found

![image](https://github.com/ExtremeFiretop/MerlinAutoUpdate-Router/assets/1971404/f9018841-e656-4af1-a6bc-08d675eba5d5)

It also falls back to the old method if no matching date is found:

![image](https://github.com/ExtremeFiretop/MerlinAutoUpdate-Router/assets/1971404/422ad357-ccfb-4341-a7a9-dac9f81b67f9)

ExtremeFiretop commented:
@Martinski4GitHub 

This is ready for a re-test :)

Martinski4GitHub commented:
> @Martinski4GitHub
> 
> This is ready for a re-test :)
>
Retesting now...

ExtremeFiretop commented:
> > @Martinski4GitHub
> > This is ready for a re-test :)
> 
> Retesting now...

Don't be shy! There was some debugging involved as you know, but last I checked this logic was working to properly estimate the runtime factoring postpone and dst and cron.

This is probably what gave me a headache yesterday lol!

ExtremeFiretop commented:
Now I should add that both Viktor and Nlurker both requested a way to view this estimate without hitting option 1.

But as you can tell by the code these some amount of processing involved to crunch all the numbers and I'm not sure I want that kinda math running Everytime.

Ideally maybe it can run once a day and save it in the file and can be referenced in the menu?

Not sure the best way to go about that. Open to ideas and feedback on the design of this. Again the idea above is just spit balling ideas I'm not glued to anything about this PR the way it is.

ExtremeFiretop commented:
> Now I should add that both Viktor and Nlurker both requested a way to view this estimate without hitting option 1.
> 
> But as you can tell by the code theres some amount of processing involved to crunch all the numbers and I'm not sure I want that kinda math running Everytime.
> 
> Ideally maybe it can run once a day and save it in the file and can be referenced in the menu?
> 
> Not sure the best way to go about that. Open to ideas and feedback on the design of this. Again the idea above is just spit balling ideas I'm not glued to anything about this PR the way it is.

Or maybe it's run once an update is found and it populates the notification date?

Martinski4GitHub commented:
@ExtremeFiretop,

I found a problem when the postponed number of days has already been reached on the days the cron jobs are supposed to run. The calculation gives me a date/time that's already gone: "`Sunday, 2024-May-05 03:10 AM`" and does not proceed with the F/W update as expected.

![RT-AC86U_FWUpdateExpected3](https://github.com/ExtremeFiretop/MerlinAutoUpdate-Router/assets/119833648/90160e20-20f9-47f3-b1e0-0d484b85df30)

The previous "old" behavior was that when the postponed days had already passed, and I selected the "Run F/W Update Check Now" option, the code went ahead and continued with the update, essentially ignoring the cron job schedule (as expected) because no postponement is in effect at that time. Now, the code stops with the already passed cron scheduled date/time.

ExtremeFiretop commented:
I understand this bit below:

> The previous "old" behavior was that when the postponed days had already passed, and I selected the "Run F/W Update Check Now" option, the code went ahead and continued with the update, essentially ignoring the cron job schedule (as expected) because no postponement is in effect at that time. Now, the code stops with the already passed cron scheduled date/time.

But I dont understand this:

> I found a problem when the postponed number of days has already been reached on the days the cron jobs are supposed to run. The calculation gives me a date/time that's already gone: "`Sunday, 2024-May-05 03:10 AM`" and does not proceed with the F/W update as expected.
> 

Can you clarify what you mean by: "The calculation gives me a date/time that's already gone: "`Sunday, 2024-May-05 03:10 AM`" and does not proceed with the F/W update as expected."

That date isn't past for me yet even so not sure how it's past/gone for you?

Martinski4GitHub commented:
> > Now I should add that both Viktor and Nlurker both requested a way to view this estimate without hitting option 1.
> > But as you can tell by the code theres some amount of processing involved to crunch all the numbers and I'm not sure I want that kinda math running Everytime.
> > Ideally maybe it can run once a day and save it in the file and can be referenced in the menu?
> > Not sure the best way to go about that. Open to ideas and feedback on the design of this. Again the idea above is just spit balling ideas I'm not glued to anything about this PR the way it is.
> 
> Or maybe it's run once an update is found and it populates the notification date?
>
I'd suggest running the full calculation **only if** one of the following events happens:

1) User changes the number of postponed days.
2) User changes the cron job schedule.
3) The notification date changes due to a new F/W Update.

For each of the above events, you make the calculation and save the value in the settings file:
For example:
`FW_New_Update_Expected_Run="2024-May-12 03:10:00"`

Then for the UI, you retrieve the value and format it for a user-friendly display:
`F/W Update expected on: Sunday, 2024-May-12 03:10 AM`

Martinski4GitHub commented:
> I understand this bit below:
> 
> > The previous "old" behavior was that when the postponed days had already passed, and I selected the "Run F/W Update Check Now" option, the code went ahead and continued with the update, essentially ignoring the cron job schedule (as expected) because no postponement is in effect at that time. Now, the code stops with the already passed cron scheduled date/time.
> 
> But I dont understand this:
> 
> > I found a problem when the postponed number of days has already been reached on the days the cron jobs are supposed to run. The calculation gives me a date/time that's already gone: "`Sunday, 2024-May-05 03:10 AM`" and does not proceed with the F/W update as expected.
> 
> Can you clarify what you mean by: "The calculation gives me a date/time that's already gone: "`Sunday, 2024-May-05 03:10 AM`" and does not proceed with the F/W update as expected."
> 
> That date isn't past for me yet even so not sure how it's past/gone for you?
>
Today **3:10 AM** has already passed!!! It's Sunday evening here!!

ExtremeFiretop commented:
> > I understand this bit below:
> > > The previous "old" behavior was that when the postponed days had already passed, and I selected the "Run F/W Update Check Now" option, the code went ahead and continued with the update, essentially ignoring the cron job schedule (as expected) because no postponement is in effect at that time. Now, the code stops with the already passed cron scheduled date/time.
> > 
> > 
> > But I dont understand this:
> > > I found a problem when the postponed number of days has already been reached on the days the cron jobs are supposed to run. The calculation gives me a date/time that's already gone: "`Sunday, 2024-May-05 03:10 AM`" and does not proceed with the F/W update as expected.
> > 
> > 
> > Can you clarify what you mean by: "The calculation gives me a date/time that's already gone: "`Sunday, 2024-May-05 03:10 AM`" and does not proceed with the F/W update as expected."
> > That date isn't past for me yet even so not sure how it's past/gone for you?
> 
> Today **3:10 AM** has already passed!!! It's Sunday evening here!!

Right sorry brain fart. Interesting okay so basically the problem is the fact that now it only does the calculation when it was supposed to run? And doesn't actually proceed with the update?

I'll be back at my desk in 10 minutes just across the sreet grabbing a snack

Martinski4GitHub commented:
> > 
> > 
> > Today **3:10 AM** has already passed!!! It's Sunday evening here!!
> 
> Right sorry brain fart. Interesting okay so basically the problem is the fact that now it only does the calculation when it was supposed to run? And doesn't actually proceed with the update?
> 
Essentially, yes.

> I'll be back at my desk in 10 minutes just across the sreet grabbing a snack
>
I've been called for dinner myself just now. I'll check back in about 1 hour.

ExtremeFiretop commented:
> > > Today **3:10 AM** has already passed!!! It's Sunday evening here!!
> > 
> > 
> > Right sorry brain fart. Interesting okay so basically the problem is the fact that now it only does the calculation when it was supposed to run? And doesn't actually proceed with the update?
> 
> Essentially, yes.
> 
> > I'll be back at my desk in 10 minutes just across the sreet grabbing a snack
> 
> I've been called for dinner myself just now. I'll check back in about 1 hour.

I just sat down now and was able to replicate your issue:

![image](https://github.com/ExtremeFiretop/MerlinAutoUpdate-Router/assets/1971404/285de244-b022-4242-987c-1e09f4ab9c38)

Will look into it a bit now :)

ExtremeFiretop commented:
I'm missing the logic from this bit:

   if [ "$((currentTimeSecs - notifyTimeSecs))" -ge "$postponeTimeSecs" ]
   then return 0 ; fi

ExtremeFiretop commented:
Fixed :)

ExtremeFiretop commented:
Almost there:

![image](https://github.com/ExtremeFiretop/MerlinAutoUpdate-Router/assets/1971404/d50a5659-3047-450d-9101-40a0eb254db3)

![image](https://github.com/ExtremeFiretop/MerlinAutoUpdate-Router/assets/1971404/4478d926-7147-4e1f-bb03-8f326d9011f1)

Martinski4GitHub commented:
> I'm missing the logic from this bit:
> 
> if [ "$((currentTimeSecs - notifyTimeSecs))" -ge "$postponeTimeSecs" ] 
> then return 0 ; fi
>
Ah, that explains it. That code is in the "old" version, so I guess you probably deleted it by mistake while making the recent changes.

ExtremeFiretop commented:
@Martinski4GitHub 

I'm done now.
Give this one more test, I think I addressed all your points, specifically these locations to run again:

User changes the number of postponed days.
User changes the cron job schedule.
The notification date changes due to a new F/W Update.

Again, don't be shy, nitpick this, we DO want to give an improved experience, but we DON"T want to break anything and that's what I'm trying to do by going slowly tonight.

ExtremeFiretop commented:
> > I'm missing the logic from this bit:
> > if [ "$((currentTimeSecs - notifyTimeSecs))" -ge "$postponeTimeSecs" ]
> > then return 0 ; fi
> 
> Ah, that explains it. That code is in the "old" version, so I guess you probably deleted it by mistake while making the recent changes.

Wait nvm, I'm not done I deleted it again...

Martinski4GitHub commented:
> Almost there:
>  ...
> 
The current title "**F/W Expected Runtime**" looks wrong to me. It's the "**F/W Update**" that's going to be run at the expected/calculated time. Something like "**F/W Update Expected Runtime**:" perhaps?  I know it's rather long but I am not sure how to shorten it without conveying the intended meaning. 

How about placing the new line at the bottom of that section (right below "USB Storage Connected")?

My 2 cents.

ExtremeFiretop commented:
Okay I am going to re-test this top to bottom and after that, that's it for me :)

ExtremeFiretop commented:
> > Almost there:
> > ...
> 
> The current title "**F/W Expected Runtime**" looks wrong to me. It's the "**F/W Update**" that's going to be run at the expected/calculated time. Something like "**F/W Update Expected Runtime**:" perhaps? I know it's rather long but I am not sure how to shorten it without conveying the intended meaning.
> 
> How about placing the new line at the bottom of that section (right below "USB Storage Connected")?
> 
> My 2 cents.

Yeah I know what you mean, I'm still thinking of a way to make it look nicer.
I do agree it's a bit long lol! Maybe it's own line would look best, still thinking on that one.

Martinski4GitHub commented:
> > > I'm missing the logic from this bit:
> > > if [ "$((currentTimeSecs - notifyTimeSecs))" -ge "$postponeTimeSecs" ]
> > > then return 0 ; fi
> > 
> > 
> > Ah, that explains it. That code is in the "old" version, so I guess you probably deleted it by mistake while making the recent changes.
> 
> Wait nvm, I'm not done I deleted it again...
>
Slow down, buddy. There's no rush, no pressure, no urgency to make it into a new release. Currently, we have no bugs that must be fixed right away. Take your time.

ExtremeFiretop commented:
> > > > I'm missing the logic from this bit:
> > > > if [ "$((currentTimeSecs - notifyTimeSecs))" -ge "$postponeTimeSecs" ]
> > > > then return 0 ; fi
> > > 
> > > 
> > > Ah, that explains it. That code is in the "old" version, so I guess you probably deleted it by mistake while making the recent changes.
> > 
> > 
> > Wait nvm, I'm not done I deleted it again...
> 
> Slow down, buddy. There's no rush, no pressure, no urgency to make it into a new release. Currently, we have no bugs that must be fixed right away. Take your time.

Yeah I'm gonna have to take a break and come back at this tomorrow.

Martinski4GitHub commented:
@ExtremeFiretop,

When I run the script with the latest code for the 1st time, it doesn't automatically calculate the F/W Update expected run time. It looks like I must update one of the related settings to trigger the calculation, is this the expected behavior for first-time executions?

![RT-AC86U_FWUpdateExpected1](https://github.com/ExtremeFiretop/MerlinAutoUpdate-Router/assets/119833648/0cf249e7-db7f-4d28-9903-0d4f0a508ae9)

I think the first time the updated script is run, it should calculate the expected date/time if the value is not yet found in the settings file. Most users will likely not know why it says "NOT SET" given that all the required values are already set & available by default (i.e. cron schedule & postponed days).

ExtremeFiretop commented:
> @ExtremeFiretop,
> 
> When I run the script with the latest code for the 1st time, it doesn't automatically calculate the F/W Update expected run time. It looks like I must update one of the related settings to trigger the calculation, is this the expected behavior for first-time executions?
> 
> ![RT-AC86U_FWUpdateExpected1](https://private-user-images.githubusercontent.com/119833648/328099626-0cf249e7-db7f-4d28-9903-0d4f0a508ae9.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQ5NzkwMzAsIm5iZiI6MTcxNDk3ODczMCwicGF0aCI6Ii8xMTk4MzM2NDgvMzI4MDk5NjI2LTBjZjI0OWU3LWRiN2YtNGQyOC05OTAzLTBkNGYwYTUwOGFlOS5qcGc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNTA2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDUwNlQwNjU4NTBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0zYTUxZmI0MzRlN2RlM2Q4MjBjMDY3NTVlMzBiY2U2MjY2NTYzZWUzMmVmY2Q1MTY2NzZjYzdiNWY2NzQ5NDQ1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.1nsn2BpPx5RNeWj8VUZCjl11qCED3oyIYU8FyUcivVI)
> 
> I think the first time the updated script is run, it should calculate the expected date/time if the value is not yet found in the settings file. Most users will likely not know why it says "NOT SET" given that all the required values are already set & available by default (i.e. cron schedule & postponed days).

Your right it probably should run the math first.
While retesting I got a bunch of errors:

![image](https://github.com/ExtremeFiretop/MerlinAutoUpdate-Router/assets/1971404/b4c63ef7-386a-491d-857d-652a1bb93773)

Clearly when I moved stuff around I missed something but my brain is shutting down it's 3AM so before I break anything more I'll reassess tomorrow.

Martinski4GitHub commented:
>
> > I think the first time the updated script is run, it should calculate the expected date/time if the value is not yet found in the settings file. Most users will likely not know why it says "NOT SET" given that all the required values are already set & available by default (i.e. cron schedule & postponed days).
> 
> Your right it probably should run the math first. While retesting I got a bunch of errors:
>
> ...
> Clearly when I moved stuff around I missed something but my brain is shutting down it's 3AM so before I break anything more I'll reassess tomorrow.
>
Yeah, you need to go to sleep & recharge Bud. By now you're running "on fumes."
You can always resume in the morning with a much clearer head & a fresh set of eyes.

Good night Bud. Sleep tight.

ExtremeFiretop commented:
> > > I think the first time the updated script is run, it should calculate the expected date/time if the value is not yet found in the settings file. Most users will likely not know why it says "NOT SET" given that all the required values are already set & available by default (i.e. cron schedule & postponed days).
> > 
> > 
> > Your right it probably should run the math first. While retesting I got a bunch of errors:
> > ...
> > Clearly when I moved stuff around I missed something but my brain is shutting down it's 3AM so before I break anything more I'll reassess tomorrow.
> 
> Yeah, you need to go to sleep & recharge Bud. By now you're running "on fumes." You can always resume in the morning with a much clearer head & a fresh set of eyes.
> 
> Good night Bud. Sleep tight.

Already in bed, I was hoping to have this all wrapped up tonight but as soon as i saw the errors I walked away.
Sleep tight buddy, I'll have this fixed up tomorrow again once I review the code with fresh eyes.

I got close, but close isn't close enough. Haha

Martinski4GitHub commented:
> 
> Already in bed, I was hoping to have this all wrapped up tonight but as soon as i saw the errors I walked away. Sleep tight buddy, I'll have this fixed up tomorrow again once I review the code with fresh eyes.
> 
> I got close, but close isn't close enough. Haha
>
That reminded me of a saying from my Father that has stuck with us (siblings) since we were kids:
"99% may be close enough but not necessarily good enough."

ExtremeFiretop commented:
> > Already in bed, I was hoping to have this all wrapped up tonight but as soon as i saw the errors I walked away. Sleep tight buddy, I'll have this fixed up tomorrow again once I review the code with fresh eyes.
> > I got close, but close isn't close enough. Haha
> 
> That reminded me of a saying from my Father that has stuck with us (siblings) since we were kids: "99% may be close enough but not necessarily good enough."

Your father sound/s/ed like a wise man. :)

ExtremeFiretop commented:
@Martinski4GitHub 

Looking back at my screenshot I just noticed the notification time updated to the same day as the script is running.
That is probably the cause of the bug/errors I am seeing. Why?

Idk my eyes are closing as I watch TV, but I didn't want to forget this tomorrow.

Martinski4GitHub commented:
> > > Already in bed, I was hoping to have this all wrapped up tonight but as soon as i saw the errors I walked away. Sleep tight buddy, I'll have this fixed up tomorrow again once I review the code with fresh eyes.
> > > I got close, but close isn't close enough. Haha
> > 
> > 
> > That reminded me of a saying from my Father that has stuck with us (siblings) since we were kids: "99% may be close enough but not necessarily good enough."
> 
> Your father sound/s/ed like a wise man. :)
>
Both my Mother & Father always encouraged us to give 100% when something was worth doing. And my Father expected us to always make the extra effort and go the extra mile. At the time, as little kids, we thought he was very strict & demanding - not in a mean way but more like he was expecting too much. But later on, sometime during the last year of high school, we slowly began to appreciate & value all his teachings, his sayings, and how he pushed us to do better and go beyond the average. Later, as adults with our own kids, we fully understood what he was doing and what he was teaching us.

Martinski4GitHub commented:
> 
> Looking back at my screenshot I just noticed the notification time updated to the same day as the script is running. That is probably the cause of the bug/errors I am seeing. Why?
> 
That's a very good test scenario. AFAIK, the built-in F/W update check runs every day sometime between 12:00 AM & 2:00 AM. This means that when a new F/W Update is found the NVRAM values will be updated at that time. Then, if our cron job is scheduled to execute every day around the same time, it could run into that same scenario where the "**Last Notification Date**" is updated the same day (and close to the same time) as the script is running.

ExtremeFiretop commented:
@Martinski4GitHub 

I think I fixed it all!!!!
Please test it and lock this in when your ready. I don't want to touch a single more thing unless I need too everything seems perfect at this moment lol!

ExtremeFiretop commented:
OKAY NOW IT'S ALL FIXED WHOOT WHOOT!
I had found more bugs. I then had to back track to find out what I changed.

Now it's perfect. No touchy from this point on for the main functions that handle the crons.
There's just so many variables to keep track of.

Only thing left is to make it run the first time if it's missing.
Also I don't think it populates anything in the Menu tab if a broken cron is used.

ExtremeFiretop commented:
That's been handled now.
It's all ready. I think with a bit of luck, this will be merged tonight :)

Fingers crossed!

ExtremeFiretop commented:
All done.

Touching up the behavior for a first run without any notification date set was a bit tricky, since all the code relies on a valid notification date lol. But as always, there's an exception to the rule that allows it to run.

Just make it believe the notification date is populated with the run date, and the rest of the logic does it's thing to add a date to the menu. It will continue to do this until a real notification date comes in.

ExtremeFiretop commented:
There is one more bug left related to setting the postpone to a delay of 2 or 3 days for some reason.

In debug mode I see this as the key difference:

```
+ date +day=%d month=%m year=%Y -d @1715299800
+ eval day=09 month=05 year=2024
+ day=09 month=05 year=2024
+ echo 05
+ sed s/^0*//
+ month=5
+ day_count=0
+ [ 0 -lt 365 ]
+ calculate_day_of_week 09 5 2024
+ day=09
+ month=5
+ year=2024
+ [ 5 -le 2 ]
+ t=15
+ K=24
+ J=20
/jffs/scripts/MerlinAU.sh: line 1: arithmetic syntax error
+ current_dow=
+ echo *
+ tr , \n
+ wc -l
+ num_dow_entries=1
+ dow_index=1
+ [ 1 -le 1 ]
+ echo *
+ cut -d, -f1
+ dow_entry=*
+ echo *
+ tr , \n
+ wc -l
+ num_month_entries=1
+ month_index=1
+ [ 1 -le 1 ]
+ echo *
+ cut -d, -f1
+ month_entry=*
+ echo *
+ sed s/^0*//
+ month_entry=*
+ [ * = * ]
+ [ * = * ]
+ [ * = * ]
```

```
+ date +day=%d month=%m year=%Y -d @1715386200
+ eval day=10 month=05 year=2024
+ day=10 month=05 year=2024
+ echo 05
+ sed s/^0*//
+ month=5
+ day_count=0
+ [ 0 -lt 365 ]
+ calculate_day_of_week 10 5 2024
+ day=10
+ month=5
+ year=2024
+ [ 5 -le 2 ]
+ t=15
+ K=24
+ J=20
+ dow=6
+ [ 6 -eq 0 ]
+ dow=5
+ echo 5
+ current_dow=5
+ echo *
+ tr , \n
+ wc -l
+ num_dow_entries=1
+ dow_index=1
+ [ 1 -le 1 ]
+ echo *
+ cut -d, -f1
+ dow_entry=*
+ echo *
+ tr , \n
+ wc -l
+ num_month_entries=1
+ month_index=1
+ [ 1 -le 1 ]
+ echo *
+ cut -d, -f1
+ month_entry=*
+ echo *
+ sed s/^0*//
+ month_entry=*
+ [ * = * ]
+ [ * = * ]
+ [ * = * ]
```

ExtremeFiretop commented:
So commit: https://github.com/ExtremeFiretop/MerlinAutoUpdate-Router/pull/205/commits/eb403af0d474f8ff5c7dcfefb54d9415b3f82aed

is the one that re-introduced the bug.
Not sure why I removed it. I seem to remember this causing me issues.
Now I have to re-test up and down again.

Seems to have fixed the issue with the postpone being set to 2 or 3 though

ExtremeFiretop commented:
@Martinski4GitHub 

I can't re-create a single bug from the days of this thread, and I've gone back through discussions and screenshots re-creating test situations. This finally has solved every issue. I've made it as clean and tight as possible.

I honestly hope you can find things to tighten, but at this point I pass this back to you. If you were to just ask to me tighten it myself I'd honestly tell you I know how to tighten many things, but this I don't know how to tighten more lol.

I've used every trick in the book you've taught me, I've used the double quotes, I've used the local and global variables, I've used the eval to make sure things run correctly and are evaluated. At this point this is as tight as I know how to make it.

I only ask that you don't completely re-do it all without good reason, because this has hurt my head to learn follow once, and at least now through debugging and trial and error I've come to get a good sense of how the general logic flows in case something else comes up.

I'm not sure I can handle learning it again from scratch a different method lol!

ExtremeFiretop commented:
> > @Martinski4GitHub
> > I can't re-create a single bug from the days of this thread, and I've gone back through discussions and screenshots re-creating test situations. This finally has solved every issue. I've made it as clean and tight as possible.
> > I honestly hope you can find things to tighten, but at this point I pass this back to you. If you were to just ask to me tighten it myself I'd honestly tell you I know how to tighten many things, but this I don't know how to tighten more lol.
> > I've used every trick in the book you've taught me, I've used the double quotes, I've used the local and global variables, I've used the eval to make sure things run correctly and are evaluated. At this point this is as tight as I know how to make it. For once if I didn't use the double quotes or local variables, it was probably for a reason lol!
> > I only ask that you don't completely re-do it all without good reason, because this has hurt my head to follow and learn once, and at least now through debugging and trial and error I've come to get a good sense of how the general logic flows in case something else comes up down the line.
> > I'm not sure I can handle learning it again from scratch a different method lol!
> 
> Well, you certainly put 110% into this, and it sounds like you've got it singing now :>). I'll review and re-test tomorrow evening. Today was a long day for me (early morning meeting + other stuff I had to do in addition to finishing up the "Lesson Learned" post-mortem report due tomorrow that goes up the ladder to my boss & my boss' boss) so right now I'm too tired to review & test any code. I'll be going to bed early tonight & will pick up again your latest versions for review & testing.
> 
> Good night buddy.

Thanks buddy! I tried. It was driving me nuts that as soon as I fixed one thing I broke another. It was a standard development wankamole.

And trust me reading back my messages I realize I sound uneasy, a d I don't mean to sound like you can't touch it . I'm just very cautious now myself since there's so many moving parts to this solution. Haha

Surely I know 30+ years of additional experience you'll find things to tighten and little things to improve but for my with my confidence and skill level, this is pretty much the end of the line for me.

Anything you touch up from here is probably a lesson learned for me though. 🤔

Martinski4GitHub commented:
@ExtremeFiretop,

The calculations of the expected F/W Update date/time are going very well; I have found no problems so far. Excellent job Bud!!

I do have one suggestion WRT the order of items listed in the top section of the Main Menu. I believe the "**F/W Product/Model ID**" should be the top item because that's the first important piece of info to show at the top. And doing that change would put "**F/W Version Installed**" & "**F/W Update Available**" items right next to each other which makes more sense because the versions info are very much related items.

I'll test a bit more later after dinner to finish my validation.

ExtremeFiretop commented:
> @ExtremeFiretop,
> 
> The calculations of the expected F/W Update date/time are going very well; I have found no problems so far. Excellent job Bud!!
> 
> I do have one suggestion WRT the order of items listed in the top section of the Main Menu. I believe the "**F/W Product/Model ID**" should be the top item because that's the first important piece of info to show at the top. And doing that change would put "**F/W Version Installed**" & "**F/W Update Available**" items right next to each other which makes more sense because the versions info are very much related items.
> 
> I'll test a bit more later after dinner to finish my validation.

No worries bud take your time! Enjoy dinner. Like you said this isn't a rush, I still want to get it finished but not mission critical.

I'm happy it looks like I locked it down for you as well! Replicated results mean great things 👍

I have no problems with your suggestions or moving the top title info! I was just playing around trying to find a way to make it look good.

Did you want to make that change or should I make that change with this PR? Let me know! 😁

ExtremeFiretop commented:
Speaking of making things look good.
Any feedback or suggestions on the wording I used around the estimate at the top?

I tried to make it as clear as possible it as short as possible.

Martinski4GitHub commented:
@ExtremeFiretop,

The following test scenarios show a date/time for "**F/W Update Expected ETA**" that is either incorrect or misleading because the scheduled cron job is not running at the time so the F/W Update cannot actually happen via a non-running cron job:

1) "**F/W Update Check**" is **DISABLED**.
   The expected date/time is misleading because no cron job is currently running since "F/W Update Check" is **DISABLED**.

![RT-AC86U_FWUpdateExpected1](https://github.com/ExtremeFiretop/MerlinAutoUpdate-Router/assets/119833648/fb304eed-4e53-497d-a2ea-a510b174b356)


2) "**F/W Update Check**" is **DISABLED** & "F/W Update available" is **NONE FOUND**.
   The expected date/time is misleading & incorrect (date/time in the past) because no cron job is currently running since "F/W Update Check" is **DISABLED**.

![RT-AX86S_FWUpdateExpected2](https://github.com/ExtremeFiretop/MerlinAutoUpdate-Router/assets/119833648/e78c38be-a896-46a9-9d42-cdb274e6ba33)

Under the above scenarios, since the cron job is not running at all, the "**F/W Update Expected ETA**" should really say "NOT SET" or perhaps "NO CRON JOB" to indicate that no update via cron job will take place. Essentially, any time "**F/W Update Check**" is DISABLED the F/W Update ETA date/time has no real meaning because no cron job is running.

Martinski4GitHub commented:
> 
> No worries bud take your time! Enjoy dinner. Like you said this isn't a rush, I still want to get it finished but not mission critical.
> 
> I'm happy it looks like I locked it down for you as well! Replicated results mean great things 👍
> 
> I have no problems with your suggestions or moving the top title info! I was just playing around trying to find a way to make it look good.
> 
> Did you want to make that change or should I make that change with this PR? Let me know! 😁
>
You can do the change now in the PR, or I can do it later. Either way would be fine with me.

Martinski4GitHub commented:
> Speaking of making things look good. Any feedback or suggestions on the wording I used around the estimate at the top?
> 
> I tried to make it as clear as possible it as short as possible.
>
Yeah, the abbreviation "**Upd**" looks weird and it seems forced. I'm still thinking of a way to word it a little better while keeping it short. I know the "Human Interface & UI Design" folks (at my day job) would reject the abbreviation based on their guidelines and standards for menu design.

Martinski4GitHub commented:
> > Speaking of making things look good. Any feedback or suggestions on the wording I used around the estimate at the top?
> > I tried to make it as clear as possible it as short as possible.
> 
> Yeah, the abbreviation "**Upd**" looks weird and it seems forced. I'm still thinking of a way to word it a little better while keeping it short. I know the "Human Interface & UI Design" folks (at my day job) would reject the abbreviation based on their guidelines and standards for menu design.

Let me sleep on it and maybe something may pop in my head.

ExtremeFiretop commented:
> @ExtremeFiretop,
> 
> The following test scenarios show a date/time for "**F/W Update Expected ETA**" that is either incorrect or misleading because the scheduled cron job is not running at the time so the F/W Update cannot actually happen via a non-running cron job:
> 
> 1. "**F/W Update Check**" is **DISABLED**.
>    The expected date/time is misleading because no cron job is currently running since "F/W Update Check" is **DISABLED**.
> 
> ![RT-AC86U_FWUpdateExpected1](https://private-user-images.githubusercontent.com/119833648/328786597-fb304eed-4e53-497d-a2ea-a510b174b356.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTUxNTI5MTMsIm5iZiI6MTcxNTE1MjYxMywicGF0aCI6Ii8xMTk4MzM2NDgvMzI4Nzg2NTk3LWZiMzA0ZWVkLTRlNTMtNDk3ZC1hMmVhLWE1MTBiMTc0YjM1Ni5qcGc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNTA4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDUwOFQwNzE2NTNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0xNGQxMzYyZmJhODYzMzNkOTAyZTcxZGU2NDhjNTllOTE0N2M3YmMzY2UzMTdhMzQwN2IxMmFiZDkxOTY0Yzk4JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.wLxJrUT9yrED3Y65YyysGK1q0rtgqAEtPTcGbztLS3E)
> 
> 2. "**F/W Update Check**" is **DISABLED** & "F/W Update available" is **NONE FOUND**.
>    The expected date/time is misleading & incorrect (date/time in the past) because no cron job is currently running since "F/W Update Check" is **DISABLED**.
> 
> ![RT-AX86S_FWUpdateExpected2](https://private-user-images.githubusercontent.com/119833648/328788638-e78c38be-a896-46a9-9d42-cdb274e6ba33.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTUxNTI5MTMsIm5iZiI6MTcxNTE1MjYxMywicGF0aCI6Ii8xMTk4MzM2NDgvMzI4Nzg4NjM4LWU3OGMzOGJlLWE4OTYtNDZhOS05ZDQyLWNkYjI3NGU2YmEzMy5qcGc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNTA4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDUwOFQwNzE2NTNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0yMGQ5ZjcwY2Q0NjMyZjc5MGM1OTRkOTdlZDY5ZDg4YjgyNTQyYWExOGE0ZDljZmNhNDIzMTE2OGQxOGRjZjI3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.XbJTDHx8d6Z_BeSi45qOegKo3I4Wcp7-BOTq6eMmFVA)
> 
> Under the above scenarios, since the cron job is not running at all, the "**F/W Update Expected ETA**" should really say "NOT SET" or perhaps "NO CRON JOB" to indicate that no update via cron job will take place. Essentially, any time "**F/W Update Check**" is DISABLED the F/W Update ETA date/time has no real meaning because no cron job is running.

Great catch!

I didn't factor in that the cron can be disabled and this information can be misleading.

I can look into both suggestions more tomorrow and see if I can address the concerns.

No issues with the calculations though? No issues with setting a postpone or cron? All the expected results were correct? It allowed you to bypass the cron if the postpone is past? Etc?

If the only thing left is this feedback I can probably poke at it tomorrow evening and get to done fairly quick I just may need to refactor some of the logic around the first run.

Martinski4GitHub commented:
>
> > Under the above scenarios, since the cron job is not running at all, the "**F/W Update Expected ETA**" should really say "NOT SET" or perhaps "NO CRON JOB" to indicate that no update via cron job will take place. Essentially, any time "**F/W Update Check**" is DISABLED the F/W Update ETA date/time has no real meaning because no cron job is running.
> 
> Great catch!
> 
> I didn't factor in that the cron can be disabled and this information can be misleading.
> 
> I can look into both suggestions more tomorrow and see if I can address the concerns.
> 
> No issues with the calculations though? No issues with setting a postpone or cron? All the expected results were correct? 
>
Yeah, the actual calculated date/time values have been correct in all my testing. I see no issues there at this point. I think the only remaining task is to account for conditions when the cron is not running so the ETA shown in the Main Menu is misleading.

>
>It allowed you to bypass the cron if the postpone is past? Etc?
> 
In this case, because the "FW Update Check" has been DISABLED since the last F/W flash, and "F/W Update Available" is NONE there's nothing to flash on that particular router. I expect things will work fine if the update check is ENABLED (that's one of my cousin's router, BTW).

ExtremeFiretop commented:
> > > Under the above scenarios, since the cron job is not running at all, the "**F/W Update Expected ETA**" should really say "NOT SET" or perhaps "NO CRON JOB" to indicate that no update via cron job will take place. Essentially, any time "**F/W Update Check**" is DISABLED the F/W Update ETA date/time has no real meaning because no cron job is running.
> > 
> > 
> > Great catch!
> > I didn't factor in that the cron can be disabled and this information can be misleading.
> > I can look into both suggestions more tomorrow and see if I can address the concerns.
> > No issues with the calculations though? No issues with setting a postpone or cron? All the expected results were correct?
> 
> Yeah, the actual calculated date/time values have been correct in all my testing. I see no issues there at this point. I think the only remaining task is to account for conditions when the cron is not running so the ETA shown in the Main Menu is misleading.
> 
> > It allowed you to bypass the cron if the postpone is past? Etc?
> 
> In this case, because the "FW Update Check" has been DISABLED since the last F/W flash, and "F/W Update Available" is NONE there's nothing to flash on that particular router. I expect things will work fine if the update check is ENABLED (that's one of my cousin's router, BTW).

I may request your help tomorrow with some of these suggestions, but I'll give it a shot tomorrow and see how clean I can make it! 

Goodnight buddy thanks again for testing and the feedback!

ExtremeFiretop commented:
Whoops wrong button. Touch screens 🤷‍♂️

Martinski4GitHub commented:
> > > > Under the above scenarios, since the cron job is not running at all, the "**F/W Update Expected ETA**" should really say "NOT SET" or perhaps "NO CRON JOB" to indicate that no update via cron job will take place. Essentially, any time "**F/W Update Check**" is DISABLED the F/W Update ETA date/time has no real meaning because no cron job is running.
> > > 
> > > 
> > > Great catch!
> > > I didn't factor in that the cron can be disabled and this information can be misleading.
> > > I can look into both suggestions more tomorrow and see if I can address the concerns.
> > > No issues with the calculations though? No issues with setting a postpone or cron? All the expected results were correct?
> > 
> > 
> > Yeah, the actual calculated date/time values have been correct in all my testing. I see no issues there at this point. I think the only remaining task is to account for conditions when the cron is not running so the ETA shown in the Main Menu is misleading.
> > > It allowed you to bypass the cron if the postpone is past? Etc?
> > 
> > 
> > In this case, because the "FW Update Check" has been DISABLED since the last F/W flash, and "F/W Update Available" is NONE there's nothing to flash on that particular router. I expect things will work fine if the update check is ENABLED (that's one of my cousin's router, BTW).
> 
> I may request your help tomorrow with some of these suggestions, but I'll give it a shot tomorrow and see how clean I can make it!
> 
> Goodnight buddy thanks again for testing and the feedback!
>
Yeah, no problem. We have each other's back on this. Later in the week, I would still like to run the latest code through our Linter tool and see what the results are. Based on that, there may be some changes to tighten up or "clean" the code a bit more, but so far it's looking good.

Sleep tight, Bud.

Martinski4GitHub commented:
@ExtremeFiretop.

I just thought of another test scenario:

3) "**F/W Update Check**" is **ENABLED** & "**F/W Update available**" is **NONE FOUND**.
Once the latest F/W image is flashed, and the F/W update available becomes NONE, giving an estimated ETA would also seem misleading because there's nothing to flash at that point.

Perhaps under all these "special" conditions, you can just hide the ETA line altogether, and show it ***only*** when the expected ETA date/time is actually meaningful (e.g. the cron job is running & there's a new F/W update to flash).

My 2 cents.

ExtremeFiretop commented:
@Martinski4GitHub 

Sorry my friend I've been slow today I just didn't have much energy or time or motivation to work on this today lol.
I figured I'd start with the easy thing in commit: https://github.com/ExtremeFiretop/MerlinAutoUpdate-Router/pull/205/commits/1da0675ab2aaf26743a618c6ac6e7159162a1d45

That addresses your top info suggestion.

The remaining suggestions summarized If I understand correctly are:

```
"F/W Update Check" is DISABLED.
The expected date/time is misleading because no cron job is currently running since "F/W Update Check" is DISABLED.

"F/W Update available" is NONE FOUND.
"F/W Update Check" is ENABLED & "F/W Update available" is NONE FOUND.
Once the latest F/W image is flashed, and the F/W update available becomes NONE, giving an estimated ETA would also seem misleading because there's nothing to flash at that point.

Under the above scenarios, since the cron job is not running at all, the "F/W Update Expected ETA" should really say "NOT SET" or perhaps "NO CRON JOB" to indicate that no update via cron job will take place. 

Perhaps under all these "special" conditions, you can just hide the ETA line altogether, and show it only when the expected ETA date/time is actually meaningful (e.g. the cron job is running & there's a new F/W update to flash).
```

Martinski4GitHub commented:
> @Martinski4GitHub
> 
> Sorry my friend I've been slow today I just didn't have much energy or time or motivation to work on this today lol. I figured I'd start with the easy thing in commit: [1da0675](https://github.com/ExtremeFiretop/MerlinAutoUpdate-Router/commit/1da0675ab2aaf26743a618c6ac6e7159162a1d45)
> 
> That addresses your top info suggestion.
> 
> The remaining suggestions summarized If I understand correctly are:
> 

I think the 2 "special" conditions under which the expected ETA is not meaningful can be succinctly summarized as follows:

1) "**F/W Update Check**" is **DISABLED**.
   Since the cron job is not running, the line for "F/W Update Expected ETA" should not be shown in the Main Menu.
   
2) "**F/W Update available**" is **NONE FOUND**
   Since there's no F/W image to flash, "F/W Update Expected ETA" should not be shown in the Main Menu.

I hope this is clear & helpful.

ExtremeFiretop commented:
@Martinski4GitHub 

Try that on for size and let me know how you feel about it to address your concerns :)
I have only tested this in the menu mode so far btw, so if you get a chance I'd test it as a cron

Martinski4GitHub commented:
> @Martinski4GitHub
> 
> Try that on for size and let me know how you feel about it to address your concerns :) I have only tested this in the menu mode so far btw, so if you get a chance I'd test it as a cron
>
I'm about to have dinner just now so I'll take a look and retest after that.

ExtremeFiretop commented:
> > @Martinski4GitHub
> > Try that on for size and let me know how you feel about it to address your concerns :) I have only tested this in the menu mode so far btw, so if you get a chance I'd test it as a cron
> 
> I'm about to have dinner just now so I'll take a look and retest after that.

Take your time I'm literally about to have dinner myself.

I know I know. It's midnight, it's the life of having a girlfriend that works evening shifts.

Sadly it can't go on for much longer.. it was nice while it lasted.

My government work is calling me back into the office 3 days a week, and has revoked my full time work from home exemption along with all other IT staff across the Government of Canada.

Means I'll have to get back on regular sleep and meal schedules and travel back into the office for no apparent reason.

Maybe you can tell I'm not impressed but I kinda knew this wasn't going to last forever.
The government has issues with IT staff recruiting and retention due to the fact that our pay is below the private sector. And the recruitments are high, sides from education and security clearances you need the dual language profile, etc. This exemption was a way to retain good staff and potentially give the GoC a leg up over the private sector.

And don't get me wrong I don't make bad pay, im easily over the average Canadian as a CS/IT-03 (you can look up the scale if your so inclined, it's all public info online)

But seems that the government has made the call that this exemption has run it's course. The  IT union is fighting but I don't have high hopes. 

All that to say sometimes starting in September I'll probably won't be up at midnight having dinner and working on this at night anymore since I'll be back in office and my girlfriend is gonna have to switch her schedules with work or find a new job.

Oh the fun life of the golden-handcuffs. There goes plans to move far away from the office and out of the city as well.

Martinski4GitHub commented:
> 
> Try that on for size and let me know how you feel about it to address your concerns :) I have only tested this in the menu mode so far btw, so if you get a chance I'd test it as a cron.
>
Excellent job buddy!! It passed all my test case scenarios. I'll merge this PR into the "dev" branch, and either Thursday or Friday evening I'll run the latest script through the Linter tool to double-check that we got "clean" code.

Martinski4GitHub reviewed (approved):
Good to go.

Martinski4GitHub commented:
> > > @Martinski4GitHub
> > > Try that on for size and let me know how you feel about it to address your concerns :) I have only tested this in the menu mode so far btw, so if you get a chance I'd test it as a cron
> > 
> > 
> > I'm about to have dinner just now so I'll take a look and retest after that.
> 
> Take your time I'm literally about to have dinner myself.
> 
> I know I know. It's midnight, it's the life of having a girlfriend that works evening shifts and working from home.
> 
> Sadly it can't go on for much longer.. it was nice while it lasted.
> 
> My government work is calling me back into the office 3 days a week, and has revoked my full time work from home exemption along with all other IT staff across the Government of Canada.
> 
> Means I'll have to get back on regular sleep and meal schedules and travel back into the office for no apparent reason.
> 
> Maybe you can tell I'm not impressed but I kinda knew this wasn't going to last forever. The government has issues with IT staff recruiting and retention due to the fact that our pay is below the private sector. And the requirements are high, sides from education and security clearances you need the dual language profile, takes years to get through the recruitment process. This exemption was a way to retain good staff and potentially give the GoC a leg up over the private sector.
> 
> And don't get me wrong I don't make bad pay, im easily over the average Canadian as a CS/IT-03 (you can look up the scale if your so inclined, it's all public info online)
> 
> But seems that the government has made the call that this exemption has run it's course. The IT union is fighting but I don't have high hopes.
> 
> All that to say sometimes starting in September I'll probably won't be up at midnight having dinner and working on this at night anymore since I'll be back in office and my girlfriend is gonna have to switch her schedules with work or find a new job.
> 
> Oh the fun life of the golden-handcuffs. There goes plans to move far away from the office and out of the city as well.
>
A similar situation has been happening here in the USA for the past 12 months or so where many employers (both private & government sectors) have been slowly terminating the "work from home" (WFH) program for pretty much all employees. However, many companies have adopted the "hybrid" schedule where some employees are allowed to work from home 2 or 3 days per week, but that's at the discretion of the immediate manager and HR. The company I work for terminated the WFH program in August 2023 and adopted a hybrid schedule but only for certain job positions and only for some categories of employees. I was lucky because I'm in the category that's allowed to choose the hybrid schedule. My wife was also lucky to be able to get a hybrid schedule for her job. But most employees were not lucky - so we cannot really complain because many other industries were hit very hard by the pandemic & the prolonged lockdowns, and people lost their jobs which they didn't get back. So yeah, we are in a golden cage.
